### ISSUE TITLE
[3] UP-DETR: Unsupervised Pre-training for Object Detection with Transformers

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#30]
## UP-DETR: Unsupervised Pre-training for Object Detection with Transformers
[paper](https://arxiv.org/abs/2011.09094) [code](https://github.com/dddzg/up-detr)

당연하게도 DETR에 대해서도 unsupervised training 을 하고자 하는 노력이 있었다.
나름 성능향상을 이끌어 냈지만 같은 시기에 나온, deformable DETR 에 비해 성능향상이 미미하여 묻힌 감이 있다. (Results section 참조)

## Random Query Patch Detection

pretraining task 를 넣으면서 2가지 문제를 풀었다.
- Multi-task learning: pretrain task 가 classification 성능을 떨어뜨리지 않게 하기 위해 **backbone을 freeze** 시키고, **patch feature reconstruction** 수행
- Multi-Query localization: 각각의 다른 object query 는 다른  area 와 boxsize 에 집중함. 이를 보여주기 위해 single-query localization 을 먼저 제시함. multi-query localization 에서는 query patch 와 object query 사이의 assignment 문제를 풀기 위해서 **object query shuffle**과 **attention mask** 를 제시.

![image](https://user-images.githubusercontent.com/16400591/136490791-40367b41-702a-4ca3-a220-c236d37b0e1f.png)
N 은 총 쿼리 수, M 은 총 random patch 수이다.
random patch를 backbone cnn + GAP 로 (2048 dim) 뽑아내고 나면 query에 더해주는 방식.
![image](https://user-images.githubusercontent.com/16400591/136491358-12590049-1e83-44d1-8ae0-a2bc48eff018.png)
DETR loss 에 `L_rec` 하나 추가된 것. reconstruction loss로는 l2를 사용함.
![image](https://user-images.githubusercontent.com/16400591/136491798-e9d5cb48-b7f1-4927-b0b5-bebce3ec2314.png)
즉, DETR output 에 그냥 2048 vector 하나 추가해 주고, regularization 하나 넣어준 것.
위에서도 말했지만, 요 녀석을 넣어서 classification 성능이 덜 떨어지게 하고 싶었다고 함. (ablation 을 보면 최종 성능에는 별로 영향 없어 보이지만, 수렴속도에서 차이가 컸다고 함)

#### Attention Mask
multi-query로 가면, query patch 간의 independence 를 보장하기 위해 attention mask를 도입.
decoder 의 self attention에 다음과 같이 mask X를 추가함
![image](https://user-images.githubusercontent.com/16400591/136492654-fc4d07d8-fe74-487e-8e4e-2c69be5faac9.png)
![image](https://user-images.githubusercontent.com/16400591/136492753-9d5d5cde-b29d-4fa7-ba8a-82560c5d66de.png)
그니까 그냥 self-attention 할 때 다른 patch는 안 보는 것.

#### Param Setup
ImageNet 1K 로 label 없이 pretraining 수행 (?!?!)
`PRETRAIN`
1. ResNet 50 BackBone은 SwAV 로 pretrain
2. shortest length [(H)320, (W)480] largest length 640 으로 resize
3. query-patch 는 128x128 로 crop, SimCLR augmentation 적용
4. AdamW 1e-4 (decay 1e-4)
5. Batch 256, 60 epoch, 

`FINETUNE`
1. CNN Backbone lr 1e-5
2. transformer lr 1e-4 
3. 나머지는 DETR과 동일

## Results
coco 2017 val 성능
![image](https://user-images.githubusercontent.com/16400591/136486727-f4180241-0470-4054-8e99-927efb4e80a0.png)
참고용으로 deformable detr 테이블도 가져왔음.
![image](https://user-images.githubusercontent.com/16400591/136486922-f65f4208-1eac-4ba7-a26c-7e45edd05906.png)

ablation. freeze backbone은 선택이 아닌 필수.
![image](https://user-images.githubusercontent.com/16400591/136492057-8933ab91-96d3-452a-a004-35b4b8e84951.png)
4가지 ablation 모델들의 학습 그래프.
![image](https://user-images.githubusercontent.com/16400591/136492304-5ca34882-0e57-4cc7-b79c-f3e740b593be.png)
